tensorflow/compiler/xla/service/plaidml/compiler.cc

```diff
@@ -327,13 +384,97 @@ StatusOr<std::unique_ptr<Program>> PlaidMLCompiler::ProgramFromHloModule (
+          // Operand 0: input tensor
+          // Operand 1: gamma
+          // Operand 2: beta
+          // Epsilon and Feature Index operands must be accessed by calling the epsilon() and feature_index() functions
+          Value findex = Value{instruction->feature_index()};
+          float e = instruction->epsilon();
+          auto batch_mean = plaidml_op::mean(instr_map[operand_ids[0]], findex);
+          auto batch_variance = plaidml_op::variance(instr_map[operand_ids[0]], findex);
+          auto batch_norm_denom = ::plaidml::edsl::sqrt(batch_variance + e);
+          auto batch_norm = ((instr_map[operand_ids[0]] - batch_mean) * instr_map[operand_ids[1]] / batch_norm_denom) + instr_map[operand_ids[2]];
+          auto tup = ::plaidml::edsl::make_tuple(batch_norm, batch_mean, batch_variance);
+          tuple_instr_map.insert(std::make_pair(cur_instr_id, tup));
+          break;
+        }
+        case HloOpcode::kConcatenate: {
+          // Tensor concatenation
+          std::vector<Tensor> tensors;
+          for (auto i = 0; i < num_operands; i++) {
+            tensors.push_back(instr_map[operand_ids[i]]);
+          }
+          int concat_axis = instruction->concatenate_dimension();
+	  auto op = plaidml_op::concatenate(tensors, concat_axis);
+          instr_map.insert(std::make_pair(cur_instr_id, op));
+          break;
+        }
         case HloOpcode::kConvolution: {
           // Tensor convolution operation
           // TODO: make sure optional operands are passed correctly
           auto op = plaidml_op::convolution(instr_map[operand_ids[0]], instr_map[operand_ids[1]]);
+          auto conv_dnums = instruction->convolution_dimension_numbers();
+          auto input_spatial_dims = conv_dnums.input_spatial_dimensions();
+          auto kernel_spatial_dims = conv_dnums.kernel_spatial_dimensions();
+          // TODO: take dims and figure out layout 
           // padding defaults to valid, test same padding and manual padding 
           // auto padding_config = instruction->padding_config();
           op = op.autopad_mode(plaidml_op::AutoPadMode::VALID);
+          // Window
+          auto raw_window = instruction->window();
+          std::vector<int> window_size;
+          std::vector<int> strides;
+          std::vector<int> dilations;
+          for (auto d : raw_window.dimensions()) {
+            if (!window_util::IsTrivialWindowDimension(d)) {
+              VLOG(2) << "Nontrivial dimension: size " << d.size() << ", stride " << d.stride() << ", base dilation " << d.base_dilation() << ", window dilation " << d.window_dilation();
+              window_size.push_back(d.size());
+              strides.push_back(d.stride());
+              dilations.push_back(d.base_dilation());
+            }
+          }
+          op = op.filter_shape(window_size);
+          op = op.strides(strides);
+          op = op.dilations(dilations);
```

Comment:
nit: there's no reason to return `op`, the fluent API is meant to be chained. But if you aren't chaining, you can just leave off the result.

So:
```
op.filter_shape(window_size)
  .strides(strides)
  .dilations(dilations)
```

And above, you can just do:
```
op.autopad_mode(plaidml_op::AutoPadMode::VALID);
```

Reply:
Done, thanks